### ISSUE TITLE
Mouse over

### ISSUE THREAD

**Issue by Dassadar** [Shoes3/shoes3#219]
Hello,

For shoes evolutions, is it planned to have a mouse over functionality compatible with stacks and flows please?
For example to display a tooltip (so that would actually be two requirements! ;-) )

Regards,
David

**Comment by passenger94** [comment]
Hello,
There is hover and leave functions on slots, you can use them to achieve some rudimentary tooltip behavior :
```ruby
Shoes.app width: 300, height: 200 do
    
    area = stack width: 50, height: 50, top: 40, left: 40 do
        background darkorange..lime
    end
    
    area.hover do |slot|
        slot.contents[0].fill = lime..darkorange
        slot.refresh_slot
        @popup = window width: 100, height: 50, title: "Tooltip" do
            inscription "Time has no respect for whatever is done without itself !"
        end unless @popup
    end
    area.leave do |slot|
        slot.contents[0].fill = darkorange..lime
        slot.refresh_slot
        @popup.close if @popup
        @popup = nil
    end

end
```
So yes there is mouse over functionalities, no nothing planned at the moment for a real tooltip/popup behavior ... 
Regards

**Comment by ccoupe** [comment]
We don't really have a light weight tool tip. Perhaps a tooltip => string style could be added to stack ,flows.  and all widgets and cursor handling could be modified detect that but that would be a very large change and very difficult to debug for all platforms.  We don't have menus and menu items for similar reasons - both are unbelievably  complicated and would change Shoes from being fun and easy.  

Request noted.  Don't hold your breath waiting.

**Comment by dredknight** [comment]
Hello everyone !

I am working with shoes for some time and really wanted to contribute! I am not a developer so I am not sure if my tips will come in handy but still here they are!

I am currently writing a small app which has 140 elements, each of which require description pop up which had to be presented next to the mouse pointer.

I tried a few things (most of them unsuccessful). At the end I reached the desired behavior with the code.
There is a hidden flow/stack named @menu. It appears where the mouse pointer is!

X and Y define the current element in array of arrays called @box, each box is 40x40 slot and may have image inside. The problem is that when empty no action should happen during interaction. When there is image there is also a hidden semi transparent bright hidden colour defined for the slot which brights the image when hovered (defined as array of arrays named @select ).

Another problem I stumbled upon is how to dynamically configure the the width and height of the @menu so the text fits. I figured out this by placing a formula which reconfigure the @menu parameters based on the length of the text (@description array).

At the end there is the toggle action which shows the pop up !

`motion do |L, T|
		@move_left=L
		@move_top=T
end


def skillz  x, y    
		@box[x][y].hover do
			unless @select[x][y]==nil then 
				@menu.style width: 210+text.length/3, height: 90+(text.length/7)*3 
				menu_left=@move_left; menu_top=@move_top;
				if @move_left+@menu.width >= app.width+10 then menu_left=@move_left-@menu.width end
				if @move_top+@menu.height >= app.height+10 then menu_top=@move_top-@menu.height end
				@menu.move(menu_left, menu_top)
				@select[x][y].toggle;
				@menu.clear do 
					background rgb(80,100,120,0.8)
					border(ghostwhite, strokewidth: 2)
					para "#{@description[x][y]}", justify: true, align: "center", stroke: white
				end
				timer(0.1) { @menu.toggle }
			end		
		end	
		@box[x][y].leave {  unless @select[x][y]==nil then @select[x][y].toggle; @menu.toggle end }
end
`

I hope this is helpful!

**Comment by passenger94** [comment]
hello dredknight, Welcome
Thank you very much!

Nice idea,
Could provide an little shoes app demonstrating the basic idea of your pop up, for example just one or two boxes with the hover/leave mechanism and of course all the minimum setting the demo might  need

for the code markdown : type this (without the quotes)  " ```ruby "  the line before your code and then the same without the word "ruby" but the line after your code
for details look up here : https://guides.github.com/features/mastering-markdown/

**Comment by dredknight** [comment]
Done! Thanks a lot!

Give me a week I will provide you the complete scenario!

**Comment by ccoupe** [comment]
@dredknight  - Welcome, We can always use help. A lot of Shoes is written in Ruby.  For this particular example, We could put up snippets/tips/sample category in the wiki and you can put/edit  you example as an article. Getting a new article linked into wiki menus is a minor pain and uploading images to the wiki is obscure.  That way your work won't be lost in a bug report.

**Comment by dredknight** [comment]
Thanks a lot! I  I am quite new to git hub so I am not really aware where should I go and write the article. I saw that I can add new page on the wiki. Is this how it is made?

**Comment by ccoupe** [comment]
@dredknight , Correct. Pick a faiirly short title  and edit away in markdown format. Save button is on the bottom. The let me know on this message what the title is and I'll setup the menu for that. (You could do that yourself too but that's a lot more effort for some just starting out. You can alway delete it you don't like and the title can be changed.

**Comment by ccoupe** [comment]
@dredknight - I took the liberty of moving your comment to https://github.com/Shoes3/shoes3/wiki/Tooltips-Fun which you can edit to your desire.

**Comment by dredknight** [comment]
Great thanks! I will wait a bit before I edit it. I want to make the method crystal clear easy first.

**Comment by passenger94** [comment]
This is fun, we should try all kind of approaches to this, shows the power of Shoes ...
@Dassadar, Building upon @dredknight idea, i tested a few things :
tooltip with shadows, a bit of thickness, transparency as @dredknight suggested, using start method to draw above all other elements, also an alternative to Shoes::Widget (as in Shoes.log or Shoes.manual)

![wiseness](https://cloud.githubusercontent.com/assets/10730930/14018128/4289a4a8-f1cc-11e5-8d74-455b79bbc225.jpg)

tiny app is here : https://gist.github.com/passenger94/e1e5c86c9fbb7a225ccf

**Comment by dredknight** [comment]
Very nice!

I am not very good at ruby yet there is one thing I do not understand in the code

```ruby
 pop_slot.instance_eval %{       ##### What instance_evail% means ? how does it work?? 
            def slide(x,y)
                move(x-#{@offset}, y-#{@offset})       ### how do you define what are you moving here?
                self
            end
            def tip=(tp)
                contents[3].text = tp           ### where is contents[] array defined?
            end
            def back_color=(pattern)
                contents[2].fill = pattern
                refresh_slot                      #### what is this? I dont see it anywhere? how come it works?
            end
        }
```

**Comment by passenger94** [comment]
Hi,
Yes that's it !
instance_eval is one of the numerous way in ruby to add methods to an object it takes either a string or a block as argument, the %{} is one of the numerous way in ruby to write a string :-D
https://en.wikibooks.org/wiki/Ruby_Programming/Syntax/Literals 
instance_eval is creating methods in the context of the receiver, here pop_slot, so in there self is that object and all calls to method are ending up on the object, for example pop_slot.contents[3]

refresh_slot is a new method recently added to slots, maybe we forget to write a manual entry !
(allows us to tell lazy Shoes to do it's work and draw immediately)

**Comment by dredknight** [comment]
Alright... I need to experiment... see more examples of this. I hope I can create some other magic with it :D

**Comment by passenger94** [comment]
look in shoes.rb and the shoes ruby library, Tons of inspiration there :-)

**Comment by dredknight** [comment]
Yes this is how I learned most of the stuff.
Just one thing do not comprehend... 

Row 60
```Ruby
 @pop.slide(@mx, @my).
```
Basically it is not defined anywhere before that and on the top of that Shoes understsands that you call SLIDE definition from the class.

Here is how I see it though not sure if correct. 

@pop  is new variable with (@)self so basically it inherits the possibility to call class instances that are in the app by default (inheritance).

If it is not like this I dont get it...

**Comment by passenger94** [comment]
Maybe there is 2 not so clear things going on here:
first remember that instance variables are available everywhere in the Shoes app block no matter where you defined them (inside the block of course) there's a chapter which talks a bit about it in the Manual : Shoes >> Rules
the second is just a matter of understanding the timing of events
i could have written the stack and events callback like this (in pseudo code) :
```ruby
main_slot = stack do
    caption ...
    @active_slot = flow ... do
        background ...
        para ...
    end
    edit_line ...
    para ...
end
main_slot.start { @pop = pop_over.slide @mx-1000, @my-1000 }

@active_slot.hover { @pop.slide(@mx, @my) ... }
@active_slot.leave { ... }
```

the @pop object is defined in the start event callback, otherwise as Shoes drawings are asynchronous, the edit_line would be drawn over the @pop ! (that's why i put that edit_line in there, just as visual feedback)
so : 
- Shoes reads the definition of the main_slot and memorize it have to draw it
- reads what's follows and learn it have to register callbacks for start, hover and leave events on some slots
- then Shoes becomes idle because it has nothing more to read, that fact triggers drawing events
- Shoes draws main_slot, when finished fire the start event of main_slot which calls the start callback

and now we have our @pop objet alive with his new methods 
you know the rest of the story :-)
hope it helps !

**Comment by dredknight** [comment]
Thanks!!! Back to work now when I have results I will share them :)

**Comment by dredknight** [comment]
I took the liberty of playing with that code you gave me.
If you change the pop_slot in the class with this thing here you get dynamically resizable pop up window. The the size depends on the length of the input string.

There are a lot of flows though. if you place a new line symbol in the string the height of the block would not be increased by one line.
there is also a problem with big words (try to place a word of 30 symbols for example). the string is not transferred on the line below when it reaches the end of the width.

It is very basic but this is as far as I could go for now. Let me know if you have any tips for how I can detect tabulation and new line in the string.

Cheers!

```Ruby
        pop_slot = stack width: @pwidth, height: 200 do
            image width: @pwidth, height: 200 do
                #rect @offset, @offset, (@pwidth-@offset*2), 100, curve: 15, fill: rgb(0,0,0,0.4), stroke: rgb(0,0,0,0)
                #shadow radius: 10, distance: 10, fill: rgb(0,0,0,0.8)
            end
            background rgb(255,255,255,0.5)
            background rgb(255,0,255,0.5)
            
            para msg 
        end
        
        pop_slot.instance_eval %{
            def slide(x,y)
                move(x-#{@offset}, y-#{@offset})
                self
            end
            def tip=(tp)
                contents[3].text = tp
				counter = tp.length
				size = 20+tp.length*10
				height=60
				while counter > 50 do
					size=400
					height+=25
					counter-=50
				end
				contents[0].style width: size , height: 200;
				#contents[0].rect.style left: #{@offset} ,top: #{@offset}, width: (size-#{@offset}),height: height 
				contents[1].style curve: 15, width: (size-#{@offset}*2-1), height: height, left: #{@offset}, top: #{@offset}
				contents[2].style curve: 15, width: (size-#{@offset}*2-3), height: height, left: #{@offset}+2, top:#{@offset}+2
				contents[3].style left: #{@offset}+10, top: #{@offset}+10, width: size-#{@offset}*2-10 
            end
            def back_color=(pattern)
                contents[2].fill = pattern
                refresh_slot
            end
        }
        
        pop_slot
```

**Comment by passenger94** [comment]
Ha ha !! Cool  ! :+1:   
I'm wondering though if you wouldn't be better to recreate completely the slot on each new text instead of trying to accommodate because it's getting real hot in there :-D
For the shadow you would have to manage the size of it also and the @offset (shadow -all effects - needs room around them otherwise you get rendering artefacts)

for a big word there is no automatic cut and line return that i'm aware of in Shoes , so you probably  have to manage this.
for tabulations and newlines, i think you would probably have to use some regex, scan the string, increase size, etc ... another level of complexity ...(maybe there is a better solution)
Looks like a candidate for an outside method, maybe an entire Class !
Also if you go for that level of complexity, maybe a module is not good anymore and a Shoes::Widget would be better ...

Meanwhile getting a clean way to do it, you could always cheat :notes: , and calculate beforehand the needed size, and feed the slot initializer with the good numbers, in most situation one provide hard coded text anyway so why not give dimensions ...

**Comment by dredknight** [comment]
passenger94 very good tips but there are some technical issues :D. 
I dont know how to create a widget.. never done such. 
Otherwise as soon as I feel more comfortable with classes I will refurbish it from top to bottom :).

**Comment by passenger94** [comment]
i just wrote something about Shoes::Widget :
https://github.com/Shoes3/shoes3/wiki/Of-Shoes::Widget-and-Shoes::Canvas

**Comment by dredknight** [comment]
Thanks this shed quite some light!
I think there is an error though.

The widget class is named CheckText but it is called  with the name of check_text. Why they are not the same?

**Comment by passenger94** [comment]
I need to make that clear, thanks !
**check_text** is a factory method : it creates an instance of **CheckText** object, the name of that method is build dynamically from the name of the Class, following established convention (not that much shared convention , have to admit ) :
https://github.com/Shoes3/shoes3/blob/master/lib/shoes.rb#L613
takes the CamelCase name of the Class, downcase it and separate 'words" delimited by Capital letters with a "_" (don't remember the word now !!)
A convenience method, so you can create the widget without much boilerplate.

**Comment by ccoupe** [comment]
> i just wrote something about Shoes::Widget :

I think I've seen my faulty memory about when Shoes.app acts like a stack. It could have been multiple  flows!  I'm easily confused! I've linked the article into the wiki menus.

**Comment by passenger94** [comment]
> It could have been multiple flows!

it traps me almost every time :-D

btw, i'm trying to collect some info about self and app not beeing the same
 if you want to add notes ...
https://gist.github.com/passenger94/833d309b8599b1e14353#file-canvas_app_and_self-md

**Comment by dredknight** [comment]
Alright guys I started doing something of a widget but kind of dont understand why is not working...

```ruby
class PopUp < Shoes::Widget
    def initialize(element, text)
		@element=element
		@text=text
		@offset=10
		motion { |x,y| @mx = x; @my = y }
        @pop_slot = stack width: 100, height: 100 do end.hide
    end	
	def create
		@element.hover do
			debug( "show" )
			@pop_slot.style width: header.length*10, height: header.length
			pop_left=@mx; 
			pop_top=@my;
			if @mx+@pop_slot.width >= app.width+@offset then pop_left=@mx-@pop_slot.width end
			if @my+@pop_slot.height >= app.height+@offset then pop_top=@my-@pop_slot.height end	
			@pop_slot.move(pop_left,pop_top)
			background rgb(248,248,255,0.8), curve: 15
			background rgb(80,100,120,0.8), curve: 15, width: @pop_slot.width-4, height: @pop_slot.height-4, left: 2, top: 2
			para "#{@text}", justify: true, align: "center", stroke: white, size: 10, width: @pop_slot.width-2*@offset, left: @offset, font: "Courier"
			@pop_slot.toggle
		end
		@element.leave {@pop_slot.toggle; debug( "close" )}
	end
	
end

Shoes.app width: 650, height: 200 do
	@h_object = flow width: 200, height:200 do
		background green
	end
	@description="tool"
    pop_up.new(@h_object, "#{@description}").create
    Shoes.show_log
end

```
I get wrong number of arguments (0 of 2) error...
I bet it is something really stupid... What I did wrong?

**Comment by passenger94** [comment]
**pop_up.new** is the faulty
you don't call new on a Widget object to initialize it, you call directly the convenience method **pop_up**   
It's a factory method ! so `pop_up(@h_object, "#{@description}").create` will do.
you might have another problem later, what is _header_ ? probably just a part of code you don't show here ?

**Comment by dredknight** [comment]
Thanks ! I took this from another code and reshaped it just for the purpose of the popup widget.

Header is actually the length of the imported text. Check appropriate changes below!
I will test and let you know.

```Ruby
class PopUp < Shoes::Widget
    def initialize(element, text)
        @element=element
        @text=text
        @offset=10
        motion { |x,y| @mx = x; @my = y }
        @pop_slot = stack width: 100, height: 100 do end.hide
    end 
    def create
        @element.hover do
            debug( "show" )
            @pop_slot.style width: @text.length*10, height: @text.length
            pop_left=@mx; 
            pop_top=@my;
            if @mx+@pop_slot.width >= app.width+@offset then pop_left=@mx-@pop_slot.width end
            if @my+@pop_slot.height >= app.height+@offset then pop_top=@my-@pop_slot.height end 
            @pop_slot.move(pop_left,pop_top)
            background rgb(248,248,255,0.8), curve: 15
            background rgb(80,100,120,0.8), curve: 15, width: @pop_slot.width-4, height: @pop_slot.height-4, left: 2, top: 2
            para "#{@text}", justify: true, align: "center", stroke: white, size: 10, width: @pop_slot.width-2*@offset, left: @offset, font: "Courier"
            @pop_slot.toggle
        end
        @element.leave {@pop_slot.toggle; debug( "close" )}
    end

end

Shoes.app width: 650, height: 200 do
    @h_object = flow width: 200, height:200 do
        background green
    end
    @description="tool"
    pop_up(@h_object, "#{@description}").create
    Shoes.show_log
**end**
```

**Comment by dredknight** [comment]
Good news!
I think it is going to work very nice :).

I just cannot understand one thing - I try to create a new stack/flow every time a pop up appears.
Unfortunately it does not seem to be removed when I use finish or remove commands.

Here is a sample code from the one above.
```ruby
class PopUp < Shoes::Widget
    def initialize(element, text)
        @element=element
        @text=text
        @offset=10
        motion { |x,y| @mx = x; @my = y }
    end 
    def create
        @element.hover do
			@pop_slot = stack do end.start
            debug( "show" )
            @pop_slot.style width: @text.length*10, height: @text.length
            pop_left=@mx; 
            pop_top=@my;
            if @mx+@pop_slot.width >= app.width+@offset then pop_left=@mx-@pop_slot.width end
            if @my+@pop_slot.height >= app.height+@offset then pop_top=@my-@pop_slot.height end 
            @pop_slot.move(pop_left,pop_top)
            background rgb(248,248,255,0.8), curve: 15
            background rgb(80,100,120,0.8), curve: 15, width: @pop_slot.width-4, height: @pop_slot.height-4, left: 2, top: 2
            para "#{@text}", justify: true, align: "center", stroke: white, size: 10, width: @pop_slot.width-2*@offset, left: @offset, font: "Courier"
            @pop_slot.toggle
        end
        @element.leave {@pop_slot.finish; debug( "close" )}
    end

end

Shoes.app width: 600, height: 600 do
    @h_object = flow width: 200, height:200 do
        background green
    end
    @description="tool"
    pop_up(@h_object, "#{@description}").create
    Shoes.show_log
end
```

**Comment by passenger94** [comment]
Great !
some remarks :
All events related methods, like start and finish, expect a block as parameter, they are not meant to be used alone, you must provide a block, inside that block you describe what you want to happen when Shoes fires that event.
So for example @pop_slot.finish won't do anything because you didn't tell Shoes what to do at the finish event of @pop_slot (that event happens when @pop_slot is removed).
Likewise for start event, it's not a command to start @pop_slot, there Shoes expects you to tell it what to do when start event happens (i.e. when the slot is drawn). In short do as you do with hover and leave.

You have to put what will be inside your pop_up (backgrounds; para) into @pop_slot or append them later

do you need the create method ? otherwise you can make everything happening in initialize   

one can also initialize the slot, move it outside visible area and `@po_slot.clear { rebuild the inners of the pop_up }` and bring it back at hover event.

Not sure to understand : do you want your pop_up to appear above the element (green area here) or below like it is in your example ?

adapting (one possible way) your code, this works for me :
```ruby
class PopUp < Shoes::Widget
    def initialize(element, text)
        @element=element
        @text=text
        @width = @text.length*12
        @height = @text.length * 8
        @offset=10
        motion { |x,y| @mx = x; @my = y }
    end 
    def create
        @element.hover do
            @pop_slot = stack left: 0, top: 0, width: @width, height: @height do 
                background rgb(248,248,255,0.8), curve: 15
                background rgb(80,100,120,0.8), curve: 15, width: @width-4, height: @height-4, left: 2, top: 2
                para "#{@text}", justify: true, align: "center", stroke: white, size: 10, width: @width-2*@offset, left: @offset, font: "Courier"
            end
            debug( "show" )
            pop_left=@mx; 
            pop_top=@my;
            if @mx+@width >= app.width+@offset then pop_left=@mx-@width end
            if @my+@height >= app.height+@offset then pop_top=@my-@height end 
            @pop_slot.move(pop_left,pop_top)
        end
        @element.leave {@pop_slot.remove; debug( "close" )}
    end

end

Shoes.app width: 600, height: 600 do
    @h_object = flow width: 200, height:200 do
        background green
    end
    @description="tool"
    pop_up(@h_object, "#{@description}").create
    Shoes.show_log
end
```

**Comment by dredknight** [comment]
hahaha thanks @passenger94 :)
The funny thing is that I have totally misunderstood start and finish.
I believed that they start and finish the event , i.e. stack.finish will do the same as stack.remove, but it was the other way around... when one task finishes (in the current example the stack) start a another one (the one in brackets).
Point taken!

I always want the popup to be on top on all stacks but I assumed this will be the case by default because when I call the class instance it creates slot and the latest created slot is always on top.

**Comment by dredknight** [comment]
check this one now. The text box is recalibrating perfectly based on text :)
```Ruby
class PopUp < Shoes::Widget
    def initialize(element, text)
        @element=element
        @text=text
		@width = @text.length*10
		@height = 50
		while @width >250 do
			case @width
				when 501..Float::INFINITY then @height+=21; @width-=250
				when 250..500 then @height+=21; @width=250
			end
		end
		
        @offset=10
        motion { |x,y| @mx = x; @my = y }
    end 
    def create
        @element.hover do
			@pop_slot = stack left: 0, top: 0, width: @width, height: @height do
				background rgb(248,248,255,0.8), curve: 15
                background rgb(80,100,120,0.8), curve: 15, width: @width-4, height: @height-4, left: 2, top: 2
                para "#{@text}", justify: true, align: "center", stroke: white, size: 10, width: @width-2*@offset, left: @offset, font: "Courier"
            end
            debug( "show" )
            pop_left=@mx; 
            pop_top=@my;
            if @mx+@width >= app.width+@offset then pop_left=@mx-@width end
            if @my+@height >= app.height+@offset then pop_top=@my-@height end 
            @pop_slot.move(pop_left,pop_top)
           # @pop_slot.toggle
        end
        @element.leave {@pop_slot.remove; debug( "close" )}
    end

end

Shoes.app width: 600, height: 600 do
    @h_object = flow width: 200, height:200 do
        background green
    end
    @description="tool ddddddddddd dddddddddddddddddddddddd  ddddddddddddv  dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd  dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd ddddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd wwwwww dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd wwwwww dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd dddddd wwwwww dddddd dddddd dddddd qqqqqq wwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwwww wwwwwwwwwwww   weeeeeeeeweeeeeeeeweeeeeeee  weeeeeeee "
    pop_up(@h_object, "#{@description}").create
    Shoes.show_log
end
```

**Comment by dredknight** [comment]
Just poppin in to say that I am still working on the popup.
I am closing on to something I appreciate more and more.
When I think it is ready I will share it here :)

**Comment by dredknight** [comment]
Hey everyone,

I bumped into a technological issue while working on the popup thing.
Lately I am very glad on how the methods are working so I decided to move to the next step and merge them into a little class.

The problem is purely connected to my inability to understand ruby classes.. I just cant get the hang out of this. Let me show you what happens exactly. I have this modules defined.

```Ruby
def set_popup img, wide, high, header, text = "", text2 = ""
		img.hover { open_popup wide, high, header, text, text2; }		
		img.leave {	close_popup }
		motion { |x, y| @move_left = x; @move_top = y }
end	
			
def open_popup wide, high, header = "", text = "", text2 = ""
		offset = 10
		@menu.style width: wide, height: high
		menu_left = @move_left; menu_top = @move_top;
		@move_left+@menu.width >= app.width+offset ? menu_left = @move_left-@menu.width : nil
		@move_top+@menu.height >= app.height+offset ? menu_top = @move_top-@menu.height : nil			
		@menu.move(menu_left, menu_top)
		@menu.clear do 
			background rgb(120,42,5,0.5), curve: 15
			background rgb(180,150,110,0.7), curve: 15, width: @menu.width-4, height: @menu.height-4, left: 2, top: 2
			header != "" ? ( tagline strong("#{header}"), stroke: white ) : nil
			text != "" ?   ( para "#{text}", justify: true, align: "center", stroke: white, size: 10, width: @menu.width-2*offset, left: offset ) : nil
			text2 != "" ?  (para "#{text2}", justify: true, align: "center", stroke: rgb(50,50,50), size: 10, width: @menu.width-2*offset, left: offset ) : nil
		end
		timer(0.1) { @menu.show }		
end
	
def close_popup
	@menu.hide
end
```

Basically when you have them you can define popup on pretty much everything with an ease.  The only drawback that I see is that one have to define the @menu (the variable that hold the hidden window that represents the popup) in the main app so it can hover globally for the whole APP and not just the insides of specific stack/flow.

Here is an example.
```Ruby
Shoes.app do
     Object1 = image "shoes.png"
     Object2 = image "gloves.png"
     set_popup ( Object1, 200, 50, "", "these are shoes" )
     set_popup ( Object2, 200, 50, "", "these are gloves" )
     @menu = stack height: 10 do end.hide
end
```
You can see how this operate in a complicated way by checking [this app](https://www.dropbox.com/s/k7pca6ohxyy03d0/skillwheel70.exe?dl=0).

Now what I want to do is migrate this functionality in Class.
I imagine this working the following way.

```Ruby
class popup
    initialize 
       @menu = stack height: 10 do end.hide  ##this stack should appear globally
    end
    
    create (object, wide, high, text)
    object.hover do
         #settting text
          # setup menu width/height
          #menu appears
    end

   object.leave do
           @menu.hide
   end
end

Shoes.app do
    pop1 = popup.new

     some_image1 = "shoes.png"
     some_image2 = "gloves.png"

     popup.create some_image1, 100, 50, "these are shoes"
     popup.create some_image2, 100, 50, "these are gloves"
end

Is it possible?

I dont know where to begin. I have basic understanding of class but it seems what class is, is not what i want while what I want is something classes offer - keeping/providing all variables in one instance.

**Comment by ccoupe** [comment]
>Is it possible?
. 
It probably is but be careful and understand where  _Shoes only Looks Like Ruby_ (tm) because it's very confusing to me too when I poke around that part of Shoes.  There's information scattered about in the manual and wiki about some of what you want. ` @menu` is a class variable of what class? Are you sure? The answer might surprise you once you get ruby to tell you.  You do need to learn how to do introspection in Ruby and understand how classes and module work before digging into what Shoes does to confuse things. Hint - read the doc for Object, Class and Module or a good book on Ruby.

If all the 'things' with a popup are of the same ' type'  you might look at subclassing shoes.widget. Look at c:"Program Files (x86)\shoes\samples\expert-custom-list-box.rb to get a feel for what can be done. It might  be the best choice or maybe not.  It deserves the 'expert' rating because it's expert territory and I'm in that group of Ruby campers.

**Comment by passenger94** [comment]
agree ! looks like a good candidate for a Shoes::Widget
otherwise you need a reference to the app object inside the class (at initialize) and explicitly call shoes methods on it, like stack :
```ruby
def initialize(theapp)
  @app = theapp
  @menu = @app.stack
end
```
you create your class like this 
```ruby
Shoes.app do
Popup.new(self)

**Comment by dredknight** [comment]
@passenger94 I just started reading your Widget article. Actually it really sounds quite manageable now!
There is one issue I foresee but I will get back to you if it really appears :). Cheers!

**Comment by dredknight** [comment]
Alright I bumped into the issue and it seems even harder than I thought. Actually I think the way Shoes works will not allow this widget to happen. First here is the code
```Ruby 
class PopUp < Shoes::Widget
		
		def initialize(img, text, options={} )
			motion { |x, y| @move_left = x; @move_top = y };
			@img = img
			@text = text
			@wide = options[:width] || 310
			@high = options[:height] || 70+(5*(text.length+text2.length))/10
			@header = options[:hdr] || ""
			@text2 = options[:txt2] || ""
			
			@img.hover { open_popup  }		
			@img.leave { close_popup }
		end
		
		def open_popup
			 motion { |x, y| @move_left = x; @move_top = y };
			debug ("app width is #{app.width}, move left is #{@move_left}")
			offset = 10
			self.style width: @wide, height: @high
			menu_left = @move_left; menu_top = @move_top;
			@move_left+self.width >= app.width+offset ? menu_left = @move_left-self.width : nil
			@move_top+self.height >= app.height+offset ? menu_top = @move_top-self.height : nil			
			self.move(menu_left, menu_top)
			self.clear do 
				background rgb(120,42,5,0.5), curve: 15
				background rgb(180,150,110,0.7), curve: 15, width: self.width-4, height: self.height-4, left: 2, top: 2
				@header != "" ? ( tagline strong("#{@header}"), stroke: white ) : nil
				@text != "" ?   ( para "#{@text}", justify: true, align: "center", stroke: white, size: 10, width: self.width-2*offset, left: offset ) : nil
				@text2 != "" ?  (para "#{@text2}", justify: true, align: "center", stroke: rgb(50,50,50), size: 10, width: self.width-2*offset, left: offset ) : nil
			end
			timer(0.1) { self.show }		
		end
		
		def close_popup
			self.hide
		end
	end 
```

Here it is how you call it

```Ruby
pop_up object, "some text", width: 240, height: 50
```

Unfortunately if the object is nested in at least one layer of slots this automatically assign that slot as owner of the widget (self is the slot). This is not ok because if the popup window is trimmed by the borders of the slot and cannot roam freely.

My question is how to make the whole Canvas always to be the owner of the popup slot nevertheless where the object/image that calls the popup is nested?

The same root issue also causes problems with motion detection of x and y.

**Comment by passenger94** [comment]
if i understand correctly,
remember a widget is ultimately just a slot, so it's inserted in the context where you initialize it unless you explicitly told it not to do so.

i think you want to initialize your widget at the app level when all other objects are drawn (to make sure it's the last one hence above the others) - maybe create a routine that redraw your widget to be always the last one, if necessary -.
At initialisation don't give it specific object nor define hover/leave events.
Create a function called,  say  "plug", give it, as parameter, an object you want to react to said events, possibly maintain an array of all "plugged" objects, define the hover/leave events for the object

Not tested, not sure, just what's comes to mind ...

**Comment by dredknight** [comment]
Alright your idea brighten me up for sure :). Here is my tryout I hope I understood you correctly.

Unfortunately it is not working as expected. What I do is the following.
No stack/flow defined, I will use the widget slot. initialize is empty.
I initialize the widget at the end of the code so it can go on top of everything.
Inside the code I use pop_up.set function to define the thingies and move the pop up slot wherever it is required. Unfortunately it is still NOT on top of the other slots and it is trimmed :(.

Here is an example

```Ruby
class PopUp < Shoes::Widget
		
		def initialize
			motion { |x, y| @move_left = x; @move_top = y }
		end
		
		def set	(img, text, options={} ) 
			@img = img
			@text = text
			@wide = options[:width] || 310
			@high = options[:height] || 70+(5*(text.length+text2.length))/10
			@header = options[:hdr] || ""
			@text2 = options[:txt2] || ""
			@img.hover { open_popup  }		
			@img.leave { close_popup }
		end
		
		def open_popup
			debug ("app width is #{app.width}, move left is #{@move_left}")
			offset = 10
			self.style width: @wide, height: @high
			menu_left = @move_left; menu_top = @move_top;
			@move_left+self.width >= app.width+offset ? menu_left = @move_left-self.width : nil
			@move_top+self.height >= app.height+offset ? menu_top = @move_top-self.height : nil			
			self.move(menu_left, menu_top)
			self.clear do 
				background rgb(120,42,5,0.5), curve: 15
				background rgb(180,150,110,0.7), curve: 15, width: self.width-4, height: self.height-4, left: 2, top: 2
				@header != "" ? ( tagline strong("#{@header}"), stroke: white ) : nil
				@text != "" ?   ( para "#{@text}", justify: true, align: "center", stroke: white, size: 10, width: self.width-2*offset, left: offset ) : nil
				@text2 != "" ?  (para "#{@text2}", justify: true, align: "center", stroke: rgb(50,50,50), size: 10, width: self.width-2*offset, left: offset ) : nil
			end
			timer(0.1) { self.show }		
		end
		
		def close_popup
			self.hide
		end
	end 

Shoes.app do 
stack do
      flow width: 40, height: 40 do 
            img1 = image "shoes.png
            pop_up.set (img1, text, options)
      end
end
pop_up    #### here I define it so it can be on top of the slots
end
```

Unfortunately it is still trimmed. How do I redraw objects so they can remain on top? This is what I want to try next.

**Comment by passenger94** [comment]
ooops i forgot ...
drawings in Shoes are asynchronous, that essentially means that there is no guarantee about when the drawings are happening, fortunately there is the start event on slots to deal with this. (start event is fired when the slot is drawn) .   
so for the root slot, try this 
```ruby
Shoes.app do 
  stack do
      flow width: 40, height: 40 do 
            img1 = image "shoes.png
            pop_up.set (img1, text, options)
      end
  end
  start { pop_up} # or self.slot.start {pop_up} 
end
```

**Comment by dredknight** [comment]
start { pop_up } - undefined method is missing. Why is that? everything is defined.

**Comment by passenger94** [comment]
i didn't look into your code ...
you can't call pop_up.set before creating pop_up widget !!

create all your objects, than in start event callback call pop_up to actually create the widget and assign a variable to it like @menu = pop_up
Then call set like @menu.set(img1, text, options)

**Comment by dredknight** [comment]
Alright! so popup definition should be done once when the whole code is written. Like this..

```Ruby
define pop widget
Shoes.app do
        create the slots on the canvas
        create objects in the slots, all objects that are going to be hovered will be assigned as @hover_box array elements.
        initialize @menu = pop_up wdiget
       Now when the whole app is written launch a cycle and rotate through all the objects and define them hover -> @menu.set @hover_box[@]
end
```
This will work, though I hoped that one will be able to define the popups inbetween the code like this.

```Ruby
define pop widget
Shoes.app do
         Initialize @menu = pop_up
        create the left slot -> Rotate through 10 elements to create them and while creating them assign popups (@menu.set)
         create the middle slot -> Rotate through 160 elements to create them and while creating them assign popups (@menu.set)
 create the right slot -> create 2 elements and assign popup on the go (@menu.set)
end
```
May be I am jinxed or something but the first way seems counter-intuitive to me.
Unless there is some way to tell a slot " You will always be my favourite. Please stay always on top" I will go with it.

Anyway the good thing is that finally I got your idea ;).

**Comment by ccoupe** [comment]
Shoes was designed for doing easy things, easily. When you go beyond that simplicity (and you are, as have many before you) then Shoes stops becoming easy and gets demanding that you do things it's way. Can you do it your way? Probably if your willing to exert great effort .Will it be worth fighting Shoes to do it your way? Most people have hit _the wall_  will quit Shoes or abandon their project or they adapt to what work with Shoes can do.

 It's a philosophical thing. There's too much low level code inside Shoes to change things now, and too few people maintaining it to for any meaningful change in the design. It's not a good thing or a bad thing. It's just Shoes.

**Comment by dredknight** [comment]
Good :), I will do it the way meant by the force of shoes then!
Props to @passenger94 for patiently explaining everything.

**Comment by dredknight** [comment]
Hey guys here are two versions you can consider PoC.

Version A:
      Uses an external menu_slot for the actual popup. 
      The restrictions are that the widget should use a slot which is defined last, so it can appear on top of other slots. This means that the widget will be initialized last in the code just after the menu_slot is created.

```Ruby 
class PopUp < Shoes::Widget
	
	def initialize ( menu )
		motion { |x, y| @move_left = x; @move_top = y }
		@menu = menu
	end
		
	def resize (wide, high, length )
		while (((high*wide/20)/length) < 8 or wide/high > 6) do
			case wide/high
				when 0..6 then high+=20
				else high+=20; wide-=wide/2
			end
			wide < 250 ? wide = 250 : nil
		end 
		return wide, high + 10
	end
		
	def present ( text, size, font = "", stroke, just, offset, wide, high)
		unless text == "" then 
			flow margin_left: offset, width: wide, height: high do
				para ("#{text}"), size: size, stroke: stroke, font: font, align: "center", justify: just
			end
		end
	end
		
    def open ( text, options={} ) 
		offset = 10
		text2 = options[:text2] || ""
		header = options[:header] || ""			
		wide = options[:width] || text.length*8
		t_high = 20
		(options[:width].is_a? Integer) == false ? (wide, t_high = resize wide, t_high, text.length) : nil
		text2 == "" ? t2_high = 0 : t2_high = 8*20*text2.length/wide + 20
		header != "" ? h_high = 60 : h_high = 0
		high = options[:height] || t_high + h_high + t2_high			
		@menu.style width: offset + wide, height: high + offset
		menu_left = @move_left; menu_top = @move_top;
		@move_left + @menu.width >= app.width + offset ? menu_left = @move_left - @menu.width : nil
		@move_top + @menu.height >= app.height + offset ? menu_top = @move_top - 2*@menu.height/3 : nil			
		@menu.move(menu_left, menu_top)
		@menu.clear do 
			background rgb(120,42,5,0.5), curve: 15
			background rgb(180,150,110,0.7), curve: 15, width: @menu.width-4, height: @menu.height - 4, left: 2, top: 2
			present  header, 17, "Bell MT", white, false, offset, wide, h_high
			present text, 9, "Courier New", white, true, offset/2, wide, t_high
			present text2, 9, "Courier New", rgb(50,50,50), true, offset/2, wide, t2_high
		end
		timer(0.1) { @menu.show }		
	end
	
	def close
		@menu.hide
	end
end

Shoes.app do
	
	flow height: 100, width: 100 do
		border("rgb(105, 105, 105)", strokewidth: 1)
		image "shoes-icon.png", left: 1, top: 1, width: 50, height: 50
		contents[1].hover { @pops.open  "morning, sunshine, cofee, beans, qeqwe ,asdasds ,wqeqwewqe 22222222222 222222222222222 2222222222222 2222 22222222222222222" }
		contents[1].leave { @pops.close } 
		
	end
	flow height: 100, width: 100 do
		border("rgb(105, 105, 105)", strokewidth: 1)
		image "shoes-icon.png", left: 1, top: 1, width: 50, height: 50
		contents[1].hover { @pops.open  "morning, sunshine, cofee, beans, qeqwe ,asdasds ,wqeqwewqe", text2: "Lets get it on" }
		contents[1].leave { @pops.close }
	end
	flow height: 100, width: 100 do
		border("rgb(105, 105, 105)", strokewidth: 1)
		image "shoes-icon.png", left: 1, top: 1, width: 50, height: 50
		contents[1].hover { @pops.open  "morning, sunshine, cofee, beans, qeqwe ,asdasds ,wqeqwewqe", text2: "Letsad  dad getsdaf it on1231d", header: "Header!" }
		contents[1].leave { @pops.close }
	end

	@menu = stack height: 10 do end.hide
	@pops = pop_up ( @menu )
end
```
Version B:
      I found a way to override shoes order of elements by using timer. When the widget is initialized its wait X seconds so the whole code is executed and then it creates the menu slot.
      This looks far better than the one above as it can be initialized everywhere!    The only problem is that we dont know if the code will manage to be executed for X seconds so timer is not very reliable if the code one is using is very complicated. Also multiple timers in the code will make this completely useless. Also the first X seconds the popup is not available for obvious reasons.

I also tried creating a slot on the canvas like this:
```Ruby
app.slot.elements[99] = stack do end.hide
```
The idea is to create the last element so it is always on top. Unfortunately Shoes does not like this code :(.

Anyway here is a sample code of version B.

```Ruby
class PopUp < Shoes::Widget
	
	def initialize ()
		motion { |x, y| @move_left = x; @move_top = y }
		timer(2) { @menu = stack height: 0 do end.hide }
	end
		
	def resize (wide, high, length )
		while (((high*wide/20)/length) < 8 or wide/high > 6) do
			case wide/high
				when 0..6 then high+=20
				else high+=20; wide-=wide/2
			end
			wide < 250 ? wide = 250 : nil
		end 
		return wide, high + 10
	end
		
	def present ( text, size, font = "", stroke, just, offset, wide, high)
		unless text == "" then 
			flow margin_left: offset, width: wide, height: high do
				para ("#{text}"), size: size, stroke: stroke, font: font, align: "center", justify: just
			end
		end
	end
		
    def open ( text, options={} ) 
		offset = 10
		text2 = options[:text2] || ""
		header = options[:header] || ""			
		wide = options[:width] || text.length*8
		t_high = 20
		(options[:width].is_a? Integer) == false ? (wide, t_high = resize wide, t_high, text.length) : nil
		text2 == "" ? t2_high = 0 : t2_high = 8*20*text2.length/wide + 20
		header != "" ? h_high = 60 : h_high = 0
		high = options[:height] || t_high + h_high + t2_high			
		@menu.style width: offset + wide, height: high + offset
		menu_left = @move_left; menu_top = @move_top;
		@move_left + @menu.width >= app.width + offset ? menu_left = @move_left - @menu.width : nil
		@move_top + @menu.height >= app.height + offset ? menu_top = @move_top - 2*@menu.height/3 : nil			
		@menu.move(menu_left, menu_top)
		@menu.clear do 
			background rgb(120,42,5,0.5), curve: 15
			background rgb(180,150,110,0.7), curve: 15, width: @menu.width-4, height: @menu.height - 4, left: 2, top: 2
			present  header, 17, "Bell MT", white, false, offset, wide, h_high
			present text, 9, "Courier New", white, true, offset/2, wide, t_high
			present text2, 9, "Courier New", rgb(50,50,50), true, offset/2, wide, t2_high
		end
		timer(0.1) { @menu.show }		
	end
	
	def close
		@menu.hide
	end
end

Shoes.app do
	
	@pops = pop_up()
	flow height: 100, width: 100 do
		border("rgb(105, 105, 105)", strokewidth: 1)
		image "shoes-icon.png", left: 1, top: 1, width: 50, height: 50
		contents[1].hover { @pops.open  "morning, sunshine, cofee, beans, qeqwe ,asdasds ,wqeqwewqe 22222222222 222222222222222 2222222222222 2222 22222222222222222" }
		contents[1].leave { @pops.close } 
		
	end
	flow height: 100, width: 100 do
		border("rgb(105, 105, 105)", strokewidth: 1)
		image "shoes-icon.png", left: 1, top: 1, width: 50, height: 50
		contents[1].hover { @pops.open  "morning, sunshine, cofee, beans, qeqwe ,asdasds ,wqeqwewqe", text2: "Lets get it on" }
		contents[1].leave { @pops.close }
	end
	flow height: 100, width: 100 do
		border("rgb(105, 105, 105)", strokewidth: 1)
		image "shoes-icon.png", left: 1, top: 1, width: 50, height: 50
		contents[1].hover { @pops.open  "morning, sunshine, cofee, beans, qeqwe ,asdasds ,wqeqwewqe", text2: "Letsad  dad getsdaf it on1231d", header: "Header!" }
		contents[1].leave { @pops.close }
	end
end
```

How the Widget works.

User is obliged to enter 1 variable of type sting only. 
```Ruby
@pops.open text
```

User can add second descriptive text if he wishes. Text is below the first one and in different colour.
```Ruby
@pops.open text,  text2: "descriptive text" 
```
A header can be added that appears on top.
```Ruby
@pops.open text,  text2: "descriptive text" , header: "Heya"
```
The widget resizes the menu based on the size of the text. This is valid for all 3 boxes - text, text2 and header!

if you dont like how the widget resizes the box you can override it by entering own values like this.
```Ruby
@pops.open text,  text2: "descriptive text" , header: "Heya", width: 300, height: 400
```

**Comment by passenger94** [comment]
@dredknight, did you tried the start event ?
It's exactly doing internally what you are trying to do on B but in the flow of Shoes events, so it waits the necessary time for elements to be ready !

**Comment by dredknight** [comment]
@passenger94 exchanged timer with start . Works wonderfully!!! Thank you!

So this is what start do? Waits for everything and then gets executed.

Is it possible to put a couple of starts in the code? if you do that in what order are they executed?

**Comment by passenger94** [comment]
Nice!
it's possible to "chain" start events i.e. wait for a canvas to be ready and inside that wait for an internal canvas to be ready. in pseudo code
```ruby
// both @s1 and @s2 are flow or stack and @s2 is inside @s1
@s1.start {
   @s2.start {
        .... code for when @s2 is ready
    }
    ... code for when @s1 is ready
}
```
I tried this with success, but know you are pushing Shoes limits very far ...

Internally Shoes queues the drawing jobs for when it is idle, at that idle time it draws something queued (a canvas) and at the end of the actual drawing routine (in C code) it notifies Shoes with, a start event, that it has finished drawing that peculiar canvas, when you use the start method in your code you hook a callback to do stuff at that very moment.
So it's not actually waiting but reacting to the fact that the drawing is finished, which from a user point of view looks like it's waiting !

**Comment by dredknight** [comment]
I hope I wont need to do chains but it really sounds fun! :D

Btw I checked how it perform on the app that I am doing and as far as performance go there is barely any memory utilization. CPU is another thing but without a profiler I cant say how it really goes.

From the user perspective it is fast :).

**Comment by ccoupe** [comment]
This [profiler](https://github.com/Shoes3/shoes3/issues/232) is in beta

**Comment by dredknight** [comment]
I am on a crossroads when it comes to what this widget should have as options.
@ccoupe  @passenger94 what do you think should be the approach?
Here are a few examples.

1. Strip it from all additional feature and auto-scale and leave just 1 text option.
```Ruby 
box.open text: @text, background: yellow, width: 300, height: 100, font: "Vivaldi",size: 10,  stroke: black
```
Here I leave the user to be able to choose the text, its font, size and stroke. As well as the background color of the popup box. Width and height are set manually.

2.  Leave auto-scale optional
Same as above except if the user do not set width and height they will be estimated automatically.
Unfortunately the user should use the default font which will be monotype so the equation can work.

3. All feature on - add ability for multiple texts and subtexts, fonts, auto-scale options etc etc..

4. If you have ideas let me know. I may have not taken into account some things that could be important.

:)

**Comment by dredknight** [comment]
Here is a sample with just plain options. 
```Ruby


	class PopUp < Shoes::Widget
	
		def initialize ()
			motion { |x, y| @move_left = x; @move_top = y }
			start { @menu = stack height: 0 do end.hide }
		end
				
		def present ( text, size=9, font = "", stroke = white, just, offset, wide, high)
			unless text == "" then
				flow margin_left: offset, width: wide, height: high do
					para ("#{text}"), size: size, stroke: stroke, font: font, align: "center", justify: just
				end
			end
		end
		
		def open ( text, options={} )
			offset = 10
			###### These are the available options options
			wide = options[:width]
			high = options[:height]
			font = options[:font]
			stroke = options[:stroke]
			size = options[:size]
			back1 = options[:border] || rgb(120,42,5,0.5)
			back2 = options[:background] || rgb(180,150,110,0.7)
			################################################# up to here
			@menu.style width: offset + wide, height: offset + high
			menu_left = @move_left; 
			menu_top = @move_top;
			@move_left + @menu.width >= app.width + offset ? menu_left = @move_left - @menu.width : nil
			@move_top + @menu.height >= app.height + offset ? menu_top = @move_top - 2*@menu.height/3 : nil
			@menu.move(menu_left, menu_top)
			@menu.clear do
				background back1, curve: 15;
				background back2, curve: 15, width: @menu.width-4, height: @menu.height - 4, left: 2, top: 2
				present text, size, font, stroke, true, offset/2, wide, high
			end
			start { @menu.show }
		end
		
		def close
			@menu.hide
		end
	end
		
	Shoes.app(title: "Open Skillwheel v:0.70, database: HoMM 5.5 RC8b", width: 1000, height: 730, resizable: false ) do
	
		@popup = pop_up()
		def pop ( slot ) ################# Play with the popup.open settings to change behaviour of popup
			slot.hover { @popup.open "Some text is passing by", width: 300, height: 200, border: rgb(0,20,5,0.5), background: rgb(0,244,99,0.5), font: "Vivaldi", size: 20, stroke: brown }
			slot.leave { @popup.close }
		end
		pop image "menu_morrowind.png", width: 300, height: 300, left: 100, top: 100
	end
```

**Comment by passenger94** [comment]
Nice ! :-))
Works fine here ! :high_heel: :high_heel: :high_heel:

Personally, i would leave minimal options with sensible defaults elsewhere and good docs for all options
(maybe a guard against missing text option ...)

**Comment by ccoupe** [comment]
I'd probably do what works for your project and not design too far into the future.  Let others ask for features or (hopefully) contribute them as needed.

**Comment by dredknight** [comment]
Thank you both :).
@ccoupe thats funny cause what works for my project is quite more complex :D

**Comment by dredknight** [comment]
Check this one! I made all options be the same as in shoes. Also put an error message when the mandatory ones (text, width, height) are not included. Try the code below.

I was wondering is there anything like help for commands in Shoes or this function is executed by the manual.

 class PopUp < Shoes::Widget

        def initialize ()
            motion { |x, y| @move_left = x; @move_top = y }
            start { @menu = stack height: 0 do end.hide }
        end

        def present ( text, size=9, font = "", stroke = white, just, offset, wide, high)
            unless text == "" then
                flow margin_left: offset, width: wide, height: high do
                    para ("#{text}"), size: size, stroke: stroke, font: font, align: "center", justify: just
                end
            end
        end

        def open ( options={} )
            
			if options[:text].nil? || options[:width].nil? || options[:height].nil? then
				error("Missing mandatory options \"width:\", \"height:\" or \"text:\"")
			else
				offset = 10
				###### These are the available options 
				text = options[:text]
				wide = options[:width]
				high = options[:height]
				font = options[:font]
				stroke = options[:stroke]
				size = options[:size]
				back1 = options[:border] || rgb(120,42,5,0.5)
				back2 = options[:background] || rgb(180,150,110,0.7)
				################################################# up to here
				@menu.style width: offset + wide, height: offset + high
				menu_left = @move_left; 
				menu_top = @move_top;
				@move_left + @menu.width >= app.width + offset ? menu_left = @move_left - @menu.width : nil
				@move_top + @menu.height >= app.height + offset ? menu_top = @move_top - 2*@menu.height/3 : nil
				@menu.move(menu_left, menu_top)
				@menu.clear do
					background back1, curve: 15;
					background back2, curve: 15, width: @menu.width-4, height: @menu.height - 4, left: 2, top: 2
					present text, size, font, stroke, true, offset/2, wide, high
				end
				start { @menu.show }
			end
        end

        def close
            @menu.hide
        end
    end

    Shoes.app(title: "Open Skillwheel v:0.70, database: HoMM 5.5 RC8b", width: 1000, height: 730, resizable: false ) do

        @popup = pop_up()
        def pop (slot=nil, text=nil, width=nil, height=nil, border=nil, background=nil, font=nil, size=nil, stroke=nil ) ################# Play with the popup.open settings to change behaviour of popup
            slot.hover { @popup.open text: text, width: width, height: height, border: border, background: background , font: font , size: size, stroke: stroke }
            slot.leave { @popup.close }
        end
        pop ( image "menu_morrowind.png", width: 200, height: 200, left: 100, top: 100 ), "All settings", 200, 200, rgb(0,20,5,0.5), rgb(0,244,99,0.5), "Vivaldi", 20, brown
		pop ( image "menu_morrowind.png", width: 200, height: 200, left: 320, top: 100 ), "Mandatory settings only", 200, 200
		pop ( image "menu_morrowind.png", width: 200, height: 200, left: 620, top: 100 ), "just text. Error message should appear when hovered"
    end
```

**Comment by passenger94** [comment]
Shaping good !!!
just for the error message :, by using the **error** method the message is send to the Shoes console which needs to be open (Alt+/) to actually see it (likewise with **debug** or **info** methods)
maybe to get some real time visual feedback, show a "special popup" with the error message ?

or maybe it opens a new window containing whole help commands and the error message, that window would also be callable via a keypress shortcut,, kind of a help page ...

**Comment by dredknight** [comment]
Done!

```Ruby
class PopUp < Shoes::Widget

        def initialize ()
            motion { |x, y| @move_left = x; @move_top = y }
            @menu = stack height: 0 do end.hide
        end

        def present ( text, size, font, stroke, just, offset, wide, high)
                flow margin_left: offset, width: wide, height: high do
                    para ("#{text}"), size: size, stroke: stroke, font: font, align: "center", justify: just
                end
        end

        def open ( options={} )
                offset, text, wide, high = 10, options[:text], options[:width], options[:height]
				options[:text].nil? || options[:width].nil? || options[:height].nil? ? (text, wide, high = "ERROR: Missing mandatory options \"width:\", \"height:\" or \"text:\"", 300, 50) : nil
                font, stroke, size, back1, back2 = options[:font] || nil, options[:stroke] || white, options[:size] || 11, options[:border] || rgb(120,42,5,0.5), options[:background] || rgb(180,150,110,0.7)
                @menu.style width: offset + wide, height: offset + high
                menu_left, menu_top = @move_left, @move_top;
                @move_left + @menu.width >= app.width + offset ? menu_left = @move_left - @menu.width : nil
                @move_top + @menu.height >= app.height + offset ? menu_top = @move_top - 2*@menu.height/3 : nil
                @menu.move(menu_left, menu_top)
                @menu.clear do
                    background back1, curve: 15;
                    background back2, curve: 15, width: @menu.width-4, height: @menu.height - 4, left: 2, top: 2
                    present text, size, font, stroke, true, offset/2, wide, high
                end
                start { @menu.show }
        end

        def close
            @menu.hide
        end
    end

    Shoes.app(title: "Open Skillwheel v:0.70, database: HoMM 5.5 RC8b", width: 1000, height: 730, resizable: false ) do

        @popup = pop_up()
        def pop (slot=nil, text=nil, width=nil, height=nil, border=nil, background=nil, font=nil, size=nil, stroke=nil )
            slot.hover { @popup.open text: text, width: width, height: height, border: border, background: background , font: font , size: size, stroke: stroke }
            slot.leave { @popup.close }
        end
        pop ( image "menu_morrowind.png", width: 200, height: 200, left: 100, top: 100 ), "All settings", 200, 200, rgb(0,20,5,0.5), rgb(0,244,99,0.5), "Vivaldi", 20, brown
        pop ( image "menu_morrowind.png", width: 200, height: 200, left: 320, top: 100 ), "Mandatory settings only", 200, 200
        pop ( image "menu_morrowind.png", width: 200, height: 200, left: 620, top: 100 )
    end
```

**Comment by dredknight** [comment]
Hey guys I did a few minor changes on the code above. Now the popup element is called with **show** and is hidden with **hide** methods to match Shoes commands.

I also wrote a [how-to guide](https://github.com/Shoes3/shoes3/wiki/Popup-text).
Let me know if I can improve the code or the guide any further.

**Comment by dredknight** [comment]
@ccoupe @passenger94 *whistling*. :P

**Comment by ccoupe** [comment]
I've been busy with with pesky internals and know I can switch to 'documenter' mode. First of all,  you did a good job on the writeup, IMHO. Playful and progressing from easy to advanced.  Well done!

How to distribute your code or include it in Shoes is actually a much more difficult question than you might expect with lots of project maintainer issues to think about, At this point in time. I think the best way is for you to create a Ruby Gem:  You'll have  to create a git hub project for 'shoes-tooltips' (or your choice of names) and structure your source directories to be gem like. and create the minimal fluff description files for a gem. PITA. Particularly the learning part,  You probably need an account at rubygems in order to upload the gem to the world. 

That's the easiest path **for me** since it means you control your gem and project and you're not dependent on me or others to do what's right for you. Big learning curve for you. 

The next alternative is to  gem-ify your code to the minimum actually used by  Shoe for ts included gems in and expecting me to (or @passenger94 or @BackOrder ) to maintain it, a responsibility I'm not willing to take, but that's just me. 

There are other hard core  positions  but they require even more learning and commitment on your part. I'm old enough to have learned not to adopt a cute kitty that the neighbor kids brought around in a cardboard box.  

Facts are: Shoes people disappear often and they  leave behind  old code behind like a box of kittens on the doorstep.   I really do suggest creating  your own  Gem & git-hub  project. 

Gem

**Comment by dredknight** [comment]
Thanks a lot! Sorry for disturbing you. I imagined it is just something you "plug in" the Ruby code.
I will check how gems are doing!

**Comment by ccoupe** [comment]
Sadly, there is a whole bunch of infrastructure inside Shoes, both code and methodological and at the project management level and as everyone knows I'm not a slave to methodology, but others feel more strongly and they are probably correct - don't tell them I said that ;-)

We don't really have place to put code like yours into Shoes except as a gem, that users can `require` if they want to. I have something  that I'd like to write, #266 which would need a similar home as yours so I have thought about the issue and gem is the fastest method. Now that I think about it more, I have several Shoes-using , Shoes-only, github projects  that need to be made into gems. Doh! Thanks for the clue stick. A gem is a lot easier than what I was contemplating for those children.  Thanks!

As you explore Gems, remember it's just a group of files and description and some rules at the HQ.  A lot to learn for such a simple task but it makes the ruby world workable for most.

**Comment by dredknight** [comment]
Good! I will go into the gem thing. I will keep you posted once I have some progress.

**Comment by dredknight** [comment]
Alright, alright. I just bumped into bug number 1. Actually it is not a bug but just... Shoes.

Here is the deal I got 4-5 boxes. The last one drawn is hidden and used for the tooltip. So far so good.
Some of the clicks around the app redraw some of the static slots. For example.
```Ruby
slot.clear do
 #replace old pic with new one?
end
```
Something simple and nothing unusual. The problem is that the content in this box is now above the tooltip box. So when hovered the picture actually overlaps the text. 
![redraw](https://cloud.githubusercontent.com/assets/12410314/18644050/7c8efa10-7eb0-11e6-9ecb-33c04ba6b441.png)

My question is how can I redraw the tooltip box to be on top? In the following example I use the default Tooltip Widget.

**Comment by ccoupe** [comment]
```ruby
slot.clear do
  remove_tooltip if tooltip.active
 # replace  old pic with new one?
 restore_last_tooltip_state
end
```
or something similar.  As you say, Shoes doesn't have a concept of layering when it comes to slots. You could also think about hiding the pic's being added and the somehow make them visible later.

**Comment by dredknight** [comment]
Thanks Cecil,

Its all good but the Tooltip is just one flow which should be redrawn from the Canvas.

Here is an app.contents info.

[(Shoes::Types::Background),@@(Tooltip)@@, (Shoes::Types::Stack), (Shoes::Types::Image), (Shoes::Types::Image), (Shoes::Types::Stack), (Tooltip), (Shoes::Types::Stack)]

The second element is the tooltip flow which is defined like this 
```Ruby
@hovers = tooltip()
```

Now my question is how to redraw it as it is not a stack or flow so I cant do clear.
The magical thing is that first I need to change the self to the App because this redraw can happen from anywhere.
I tried this
```Ruby
@hovers = app.tooltip()
```
and this
```Ruby
app.hovers = tooltip()
```
But both are giving errors. Is it possible at all?

May be using the contents like this:
```Ruby
app.contents[1].#do something with the toolstack??##
```
But I dont know what exactly. Being searching some methods forwards and backwards but with no success so far.

**Comment by ccoupe** [comment]
Personally, i find APPS and app.method to be very confusing. It is possible to walk through it and find things once you understand it, but I forget because it makes no sense 60 minutes later. Not much help, sorry.

**Comment by dredknight** [comment]
Alright  I think I got the hang out of it. There is one thing I dont know how to cope with.
I dont understand 2 things. Why when I try to redraw the tooltip it says in debugger
     undefined method `tooltip' for (Shoes::Types::App "Shoes")

Also I fear that even if reassign the slot to the variable the SET method should be reapplied on the image. Am I correct? if the answer is yes this means to reapply all sets in the app which is completely dead end for me :(.

Here is an example.

```Ruby
class Tooltip < Shoes::Widget

    def initialize ()
         motion { |x, y| @move_left = x; @move_top = y }
         start { @@menu = stack height: 0 do end.hide }
    end

	def show( options={} )
		offset, high, wide, text, stroke, size, font, back1, back2 = 10, options[:height], options[:width], options[:text], options[:stroke], options[:size], options[:font],  options[:border] || rgb(120,42,5,0.5), options[:background] || rgb(180,150,110,0.7) 
        ( wide.nil? || high.nil? || text.nil? ) ? ( high = 250; wide = 400; size = 10; text = "ERROR: Missing mandatory options \"width:, \"height:\" or \"text:\"\nAvailable options:\ntext: the text presented in the tooltip window\nwidth: the width of the tooltip\nheight: the height of the tooltip font: the font of the text\nstroke: the stroke of the text\nsize: the size of the text\nborder: the colour of the border of the tooltip\nbackground: the colour of the tooltip box\nExample: slot.hover { @tooltip.show text: \"tooltip comming through\", width: 200, height: 100, border: rgb(0,20,5,0.5), background: rgb(0,244,99,0.5) , font: Vivaldi , size: 10, stroke: white \} "  ) : nil
        @@menu.style width: offset + wide, height: offset + high
		menu_left, menu_top = @move_left, @move_top;
		@move_left + @@menu.width >= app.width + offset ? menu_left = @move_left - @@menu.width : nil
		@move_top + @@menu.height >= app.height + offset ? menu_top = @move_top - 2*@@menu.height/3 : nil
		@@menu.move(menu_left, menu_top)
		@@menu.clear do
			background back1, curve: 15
			background back2, curve: 15, width: @@menu.width-4, height: @@menu.height - 4, left: 2, top: 2
			flow margin_left: offset, width: wide, height: high do
				para ("#{text}"), size: size, stroke: stroke, font: font, align: "center", justify: true
			end
		end
		start { @@menu.show }
	end

	def hide
		@@menu.hide
	end
end 

Shoes.app do
background green
@hovers = tooltip()

	def set  (img, options={} )
			img.hover { @hovers.show  text: options[:text], width: options[:width], height: options[:height]; }	
			img.leave { @hovers.hide; }
	end

	set ( image "#{DIR}/static/app-icon.png", width: 200, height: 200, left: 320, top: 100 ), text: "Mandatory settings only", width: 200, height: 200

	button "redraw the tooltip" do 
		   @hovers.remove
		   @hovers = tooltip()
	end
end
````

**Comment by dredknight** [comment]
Fixed!
The pop up is now stack in stack.
When the when popup is called (show option) the internal stack is redrawn each time.
Not sure how this is performance wise. The user experience is still good but I see that each redraw increase the usage of the memory by a small bit without releasing it.

I have seen this before on other parts of the app that regard redrawing a.k.a. @@slot.clear
Is there anything that can be done to free memory ? 
As far as I know ruby is handling this on its own but as Shoes is half C half ruby this is under consideration.

Here is the working class code.

```Ruby
class Tooltip < Shoes::Widget

    def initialize ()
         motion { |x, y| @move_left = x; @move_top = y }
         @@menu1 = stack;
    end

    def show( options={} )
         @@menu1.clear {	@@menu = stack height: 0 do end.hide }
        offset, high, wide, text, stroke, size, font, back1, back2 = 10, options[:height], options[:width], options[:text], options[:stroke], options[:size], options[:font],  options[:border] || rgb(120,42,5,0.5), options[:background] || rgb(180,150,110,0.7) 
        ( wide.nil? || high.nil? || text.nil? ) ? ( high = 250; wide = 400; size = 10; text = "ERROR: Missing mandatory options \"width:, \"height:\" or \"text:\"\nAvailable options:\ntext: the text presented in the tooltip window\nwidth: the width of the tooltip\nheight: the height of the tooltip font: the font of the text\nstroke: the stroke of the text\nsize: the size of the text\nborder: the colour of the border of the tooltip\nbackground: the colour of the tooltip box\nExample: slot.hover { @tooltip.show text: \"tooltip comming through\", width: 200, height: 100, border: rgb(0,20,5,0.5), background: rgb(0,244,99,0.5) , font: Vivaldi , size: 10, stroke: white \} "  ) : nil
        @@menu.style width: offset + wide, height: offset + high
        menu_left, menu_top = @move_left, @move_top;
        @move_left + @@menu.width >= app.width + offset ? menu_left = @move_left - @@menu.width : nil
        @move_top + @@menu.height >= app.height + offset ? menu_top = @move_top - 2*@@menu.height/3 : nil
        @@menu.move(menu_left, menu_top)
        @@menu.clear do
            background back1, curve: 15
            background back2, curve: 15, width: @@menu.width-4, height: @@menu.height - 4, left: 2, top: 2
            flow margin_left: offset, width: wide, height: high do
                para ("#{text}"), size: size, stroke: stroke, font: font, align: "center", justify: true
            end
        end
        start { @@menu.show }
    end

    def hide
        @@menu.hide
    end
end 

Shoes.app do
background green
@hovers = tooltip()

    def set  (img, options={} )
            img.hover { @hovers.show  text: options[:text], width: options[:width], height: options[:height]; } 
            img.leave { @hovers.hide; }
    end

    set ( image "#{DIR}/static/app-icon.png", width: 200, height: 200, left: 320, top: 100 ), text: "Mandatory settings only", width: 200, height: 200

    button "redraw the tooltip" do 
           @hovers.remove
           @hovers = tooltip()
    end
end
```

**Comment by passenger94** [comment]
Congrats, i think you find a really nasty bug !
minimal case:
```ruby
class CheckText < Shoes::Widget
    def initialize(text, active=true)
        check checked: active
        para text
    end
end

Shoes.app width: 360, height: 200 do
    #check_text "sure" <-- works fine
    button "redraw" do
        check_text "no, please", false
    end
end
```
This is puzzling me, never happened to me, never heard of it (@cecil ?) even though i have custom widgets deep buried into flows and stacks, and they do still work just fine !!
Fortunately workaround is easy just call the method on app.slot (or directly on slot here)
```ruby
    button "redraw" do
        slot.check_text "no, please", false
    end
```

as a rule of thumb, if you hit again that kind of behavior, try to send the method on slot on slot instead of app (or self)

**Comment by passenger94** [comment]
very strange, i tried your fix on the minmal example doesn't work ....

**Comment by dredknight** [comment]
Hey @passenger94  thank you for the great advice!
slot.things will help a lot. I will try it out and see if it turns up something better.

Regarding the memory note, I compared this version of the app with the old one. It seems both are gaining small amounts of memory but I saw it just now.

**Comment by passenger94** [comment]
ok i think i know what's going on
```ruby
  class Types::Widget
    @types = {}
    def self.inherited(subc)
      methc = subc.to_s[/(^|::)(\w+)$/, 2]
              .gsub(/([A-Z]+)([A-Z][a-z])/, '\1_\2')
              .gsub(/([a-z\d])([A-Z])/, '\1_\2').downcase
      @types[methc] = subc
      Shoes.class_eval %{
        def #{methc}(*a, &b)
          a.unshift Widget.instance_variable_get("@types")[#{methc.dump}]
          widget(*a, &b)
        end
      }
    end
  end
```
the widget method is created by class_eval-ing on Shoes class so the method becomes an instance methods of Shoes, which ends up being the Class of the slot of an app object ... (the canvas of the app ... confusing hé !?), under "normal" circumstances every method unknown to app is redirected on the app Canvas (slot), but not in our case at hand ... 

i quickly tried this and it seems to work (even deep buried into flows/stacks), thoughts ?
as Widgets belongs to the App level and self/app can be reach everywhere ...
class_eval-ing on Shoes::App instead
```ruby
  class Types::Widget
    @types = {}
    def self.inherited(subc)
      methc = subc.to_s[/(^|::)(\w+)$/, 2]
              .gsub(/([A-Z]+)([A-Z][a-z])/, '\1_\2')
              .gsub(/([a-z\d])([A-Z])/, '\1_\2').downcase
      @types[methc] = subc
      Shoes::Types::App.class_eval %{
        def #{methc}(*a, &b)
          a.unshift Shoes::Types::Widget.instance_variable_get("@types")[#{methc.dump}]
          widget(*a, &b)
        end
      }
    end
  end
```

**Comment by dredknight** [comment]
I am not sure where is this code from but if this is the Widget code and you fixed this will this be implemented in next version of Shoes?

Good job ;)

**Comment by ccoupe** [comment]
A fair amount of shoes magic is done in Ruby/Shoes which you can inspect\modify. That code is in lib/shoes.rb, line 459 or nearby.  Probably c:\Program Files (x86)\shoes\lib\shoes.rb  for you.

When @passenger94 says `needs more testing` that means I need to do that and then bless his changes and since I find it a very obscure part of Shoes that I don't use but you do (so it's tested), it will probably be in the next release and I've tagged the issue so I won't forget it.

**Comment by dredknight** [comment]
Rejoice!

**Comment by ccoupe** [comment]
Since I'm deep in the plot woods and 3.3.2 gets farther away, it would be best if you monkey patch (replace the existing Shoes widget creation code with you own, in your app).

**Comment by dredknight** [comment]
I will do that and test.

**Comment by BackOrder** [comment]
Will tooltips included in Shoes Samples?

**Comment by ccoupe** [comment]
Another good idea @BackOrder !  Instead of @dredknight having to weaponize his code into a gem or inventing  some odd Shoes class, just provide the souce in a working sample. I like it.

**Comment by dredknight** [comment]
This is good thing! Btw here is the latest sample of the code that is 100% reliable and tested with multiple layers of flows and stacks and is working under many different redraws of canvas slots during the app time frame.

A small example is added as well.

P.S. I have also polished the popup-auto-resize script. I will add it in the wiki some time in the future.

```Ruby
class Tooltip < Shoes::Widget

    def initialize ()
        motion { |x, y| @move_left = x; @move_top = y }
        @menu1 = stack;
    end

    def show ( options={} )
        @menu1.clear { @menu = stack height: 0 do end.hide }
        offset, high, wide, text, stroke, size, font, back1, back2 = 10, options[:height], options[:width], options[:text], options[:stroke], options[:size], options[:font],  options[:border] || rgb(120,42,5,0.5), options[:background] || rgb(180,150,110,0.7) 
        ( wide.nil? || high.nil? || text.nil? ) ? ( high = 250; wide = 400; size = 10; text = "ERROR: Missing mandatory options \"width:, \"height:\" or \"text:\"\nAvailable options:\ntext: the text presented in the tooltip window\nwidth: the width of the tooltip\nheight: the height of the tooltip font: the font of the text\nstroke: the stroke of the text\nsize: the size of the text\nborder: the colour of the border of the tooltip\nbackground: the colour of the tooltip box\nExample: slot.hover { @tooltip.show text: \"tooltip comming through\", width: 200, height: 100, border: rgb(0,20,5,0.5), background: rgb(0,244,99,0.5) , font: Vivaldi , size: 10, stroke: white \} "  ) : nil
        @menu.style width: offset + wide, height: offset + high
        menu_left, menu_top = @move_left, @move_top;
        @move_left + @menu.width >= app.width + offset ? menu_left = @move_left - @menu.width : nil
        @move_top + @menu.height >= app.height + offset ? menu_top = @move_top - 2*@menu.height/3 : nil
        @menu.move(menu_left, menu_top)
        @menu.clear do
            background back1, curve: 15
            background back2, curve: 15, width: @menu.width-4, height: @menu.height - 4, left: 2, top: 2
            flow margin_left: offset, width: wide, height: high do
                para ("#{text}"), size: size, stroke: stroke, font: font, align: "center", justify: true
            end
        end
        start { @menu.show }
    end

    def hide
        @menu.hide
    end
end 

Shoes.app do
background green
@hovers = tooltip()

    def set  (img, options={} )
			img.hover { @hovers.show options } 
            img.leave { @hovers.hide; }
    end
	set ( image "#{DIR}/static/app-icon.png", width: 50, height: 50, left: 30, top: 100 ), text: "No options"
    set ( image "#{DIR}/static/app-icon.png", width: 50, height: 50, left: 120, top: 100 ), text: "Mandatory settings only", width: 200, height: 200
	set ( image "#{DIR}/static/app-icon.png", width: 50, height: 50, left: 340, top: 100 ), text: "ALL SETTINGS", width: 200, height: 200, stroke: orange, size: 14, font: "Vivaldi", border: blue, background: yellow
end
```

**Comment by ccoupe** [comment]
I've made some tiny changes, just because.  This will be in the next beta. Thanks @dredknight 

The changes:
```ruby
Shoes.app do
background palegreen
para "Demonstrates how to write tool tips. Credit and thanks to @dredknight\n" \
 "Move the mouse to the icons below" 
@hovers = tooltip()

    def set  (img, options={} )
            img.hover { @hovers.show options } 
            img.leave { @hovers.hide; }
    end
    set ( image "#{DIR}/static/app-icon.png", width: 50, height: 50, left: 30, top: 100 ), text: "No options"
    set ( image "#{DIR}/static/app-icon.png", width: 50, height: 50, left: 120, top: 100 ), text: "Mandatory settings only", width: 200, height: 200
    set ( image "#{DIR}/static/app-icon.png", width: 50, height: 50, left: 340, top: 100 ), text: "ALL SETTINGS", width: 200, height: 200, stroke: orange, size: 14, font: "Serif", border: blue, background: yellow
end

**Comment by BackOrder** [comment]
Great job @dredknight !

Quick question: is the following screenshot normal? I'd rather expect tooltips to support autofit to any given text.

![image](https://cloud.githubusercontent.com/assets/8844578/19794654/f373b4ea-9ca1-11e6-8b51-d49a0210a811.png)

**Comment by ccoupe** [comment]
How does one measure displayed or about to be display text width in Shoes? And change it? And keep Shoes simple?

**Comment by BackOrder** [comment]
@ccoupe well, haven't we ran into such problem before? There are some runtime information unavailable to us. We should consider adding that in our roadmap or even have an entry in the issue tracker.

```ruby
Shoes.app do
	flow do
		background green
		@p = para "this is some text"
	end
	
	@f = flow do
		background pink
		@p2 = para "this is some text"
	end
	
	timer(0) { 
		para "height: #{@p.height}"
		@f.height = @p2.height * 2
	}
end
```

**Comment by BackOrder** [comment]
@ccoupe by the way, the way Shoes work with flows... it should autofit when no height is specified anyway. One has to make the decoration relative positions and that's all.

**Comment by BackOrder** [comment]
![image](https://cloud.githubusercontent.com/assets/8844578/19796093/caa77034-9caf-11e6-9bdd-f57f19c5bff4.png)


```ruby
Shoes.app(width: 250, height: 250) do
	@menu = stack width: 100, left: 75, top: 100 do end
	@menu.clear do
		@b = background blue, curve: 15
		@b2 = background yellow, curve: 15, left: 2, top: 2, height: @menu.height - 4
		flow margin_left: 12, margin_top: 5 do
			@p = para "Say hello!"
		end
		timer(0) { 
			@b.width = @p.width + 20
			@b2.width = @b.width - 4
		}
	end
end
```

**Comment by ccoupe** [comment]
I welcome all code contributions for rewriting textblock and the layout engine (be backwards compatible though) or perhaps you can design a new layout mechanism that can work in parallel with old code.  (tk's grid layout is awesome once you figure it out) and Java grid-bag layout derives from that ...

Since this issue (#219) is about tool-tips in response to a mouse hover/click , let's keep it on topic.  @Dassadar  wanted something for slots and flow. Others have asked for tool tip support for a simple  para or para segment. TextBlock is not a widget, never has been - it's a Cairo abstraction. . Gtk has a label widget if memory serves and so does OSX (I think) so you could file a new issue asking for that. Please think of the api you need for that.

**Comment by dredknight** [comment]
@BackOrder  width and height is what sets the size of the block.

I have made an engine that auto-scale the when font is monotype and for specific size.
Make a new ticket and I we can elaborate on it there.

**Comment by dredknight** [comment]
Closing this one as solved! @Dassadar if you have any questions you can reopen it or send me private message.
